Source/Core/InputCommon/ControllerInterface/WGInput/WGInput.h

```diff
@@ -0,0 +1,222 @@
+// Copyright 2018 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#pragma once
+
+#pragma warning(push)
+#pragma warning(disable : 4265)
+#include <wrl.h>
+#pragma warning(pop)
+
+#include <windows.gaming.input.h>
+
+#include "InputCommon/ControllerInterface/ControllerInterface.h"
+
+namespace ciface
+{
+namespace WGInput
+{
+namespace WGI = ABI::Windows::Gaming::Input;
+using Microsoft::WRL::ComPtr;
+
+void Init();
+void DeInit();
+void PopulateDevices();
+
+class Device : public Core::Device
+{
+protected:
+  class Button : public Core::Device::Input
+  {
+  public:
+    Button(const boolean& _button) : m_button(_button) {}
+    ControlState GetState() const override;
+
+  private:
+    const boolean& m_button;
+  };
+
+  // A button with one of the "labels" that WGI provides:
+  class NamedButton : public Button
+  {
+  public:
+    NamedButton(const boolean& _button, const char* _name) : Button(_button), m_name(_name) {}
+    std::string GetName() const override;
+
+  private:
+    const char* const m_name;
+  };
+
+  // A button with no label so we name it by its index:
+  class IndexedButton : public Button
+  {
+  public:
+    IndexedButton(const boolean& _button, u32 _index) : Button(_button), m_index(_index) {}
+    std::string GetName() const override;
+
+  private:
+    u32 m_index;
+  };
+
+  class Axis : public Core::Device::Input
+  {
+  public:
+    Axis(const double& _axis, double _base, double _range)
+        : m_base(_base), m_range(_range), m_axis(_axis)
+    {
+    }
+
+    ControlState GetState() const override;
+
+  protected:
+    const double m_base, m_range;
```

Comment:
prefer placing variables on their own line. These being const also makes the class non-movable.